### PR TITLE
Improve design of internal/stanza log converter

### PR DESCRIPTION
**Pull request by pmalek-sumo**
**Description:** Improve the internal design of internal/stanza log converter.

* get rid of locks
* rely purely on channel communication
* introduce independent units of work: `workerLoop`, `batchLoop`, `flushLoop`
* introduce configurable worker pool

**Testing:** existing unit tests pass + added a converter benchmark

### CODE REVIEW

**Comment by pmalek-sumo** [comment]
Testbed results comparison:

81dc4416: 

```
( make otelcontribcol && cd testbed/ && TESTCASE_DURATION=15s TEST_ARGS="-run TestLog -count 1" ./runtests.sh)
...
# Test PerformanceResults
Started: Thu, 15 Apr 2021 12:07:50 +0200

Test                                    |Result|Duration|CPU Avg%|CPU Max%|RAM Avg MiB|RAM Max MiB|Sent Items|Received Items|
----------------------------------------|------|-------:|-------:|-------:|----------:|----------:|---------:|-------------:|
Log10kDPS/OTLP                          |PASS  |     17s|    17.3|    14.7|         40|         49|    149900|        149900|
Log10kDPS/filelog                       |PASS  |     15s|    18.2|    19.3|         41|         49|    150000|        150000|
Log10kDPS/kubernetes_containers         |PASS  |     15s|    44.5|    45.1|         46|         54|    150000|        150000|
Log10kDPS/k8s_CRI-Containerd            |PASS  |     15s|    40.0|    40.2|         46|         54|    150000|        150000|
Log10kDPS/k8s_CRI-Containerd_no_attr_ops|PASS  |     15s|    33.0|    33.4|         45|         53|    150000|        150000|
Log10kDPS/CRI-Containerd                |PASS  |     15s|    19.6|    20.0|         44|         52|    150000|        150000|
Log10kDPS/syslog-udp                    |PASS  |     15s|    59.3|    64.0|         42|         51|    139900|        139900|
Log10kDPS/syslog-tcp-batch-1            |PASS  |     15s|    53.4|    55.5|         42|         50|    149900|        149900|
Log10kDPS/syslog-tcp-batch-100          |PASS  |     15s|    45.9|    47.6|         41|         49|    150000|        150000|
Log10kDPS/FluentBitToOTLP               |PASS  |     23s|     5.0|     8.3|         52|         70|    149900|        149900|
Log10kDPS/FluentForward-SplunkHEC       |PASS  |     16s|    35.0|    35.6|         55|         68|    148800|        148800|
```

this PR:

```
Started: Thu, 15 Apr 2021 12:17:23 +0200

Test                                    |Result|Duration|CPU Avg%|CPU Max%|RAM Avg MiB|RAM Max MiB|Sent Items|Received Items|
----------------------------------------|------|-------:|-------:|-------:|----------:|----------:|---------:|-------------:|
Log10kDPS/OTLP                          |PASS  |     17s|    15.2|    12.7|         38|         47|    149900|        149900|
Log10kDPS/filelog                       |PASS  |     15s|    20.4|    22.2|         42|         49|    150000|        150000|
Log10kDPS/kubernetes_containers         |PASS  |     15s|    45.2|    45.5|         45|         53|    150000|        150000|
Log10kDPS/k8s_CRI-Containerd            |PASS  |     15s|    43.4|    44.8|         45|         54|    150000|        150000|
Log10kDPS/k8s_CRI-Containerd_no_attr_ops|PASS  |     15s|    39.7|    41.0|         44|         52|    150000|        150000|
Log10kDPS/CRI-Containerd                |PASS  |     15s|    27.4|    28.3|         45|         53|    150000|        150000|
Log10kDPS/syslog-udp                    |PASS  |     15s|    58.3|    66.3|         43|         51|    139900|        139900|
Log10kDPS/syslog-tcp-batch-1            |PASS  |     15s|    52.5|    59.1|         41|         49|    149900|        149900|
Log10kDPS/syslog-tcp-batch-100          |PASS  |     15s|    50.7|    53.3|         42|         49|    149900|        149900|
Log10kDPS/FluentBitToOTLP               |PASS  |     23s|     4.9|     7.0|         52|         76|    149900|        149900|
Log10kDPS/FluentForward-SplunkHEC       |PASS  |     15s|    34.0|    34.7|         56|         67|    149600|        149600|
```

File: receiver/filelogreceiver/go.mod [hunk 1]
```diff
@@ -4,6 +4,7 @@ go 1.15
 
 require (
 	github.com/observiq/nanojack v0.0.0-20201106172433-343928847ebc
+	github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage v0.24.1-0.20210416221238-c731ffe72642 // indirect
```
Comment (pmalek-sumo) [inline receiver/filelogreceiver/go.mod:7]: Needed to tweak this due to:

```
receiver/filelogreceiver$ go mod download ; go mod tidy
go: finding module for package github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage/filestorage
go: downloading github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage v0.24.0
go: downloading github.com/open-telemetry/opentelemetry-collector-contrib v0.24.0
github.com/open-telemetry/opentelemetry-collector-contrib/receiver/filelogreceiver imports
        github.com/open-telemetry/opentelemetry-collector-contrib/internal/stanza tested by
        github.com/open-telemetry/opentelemetry-collector-contrib/internal/stanza.test imports
        github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage/filestorage: module github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage@latest found (v0.24.0), but does not contain package github.com/open-telemetry/opentelemetry-collector-contrib/extension/storage/filestorage
```

**Comment by pmalek-sumo** [comment]
@djaglowski Let me know whether there's any sense in me rebasing this and resolving conflicts so that we'd like to keep looking at it or reject it straight away.

**Comment by djaglowski** [comment]
@pmalek-sumo This design is certainly easier to reason about. It's unfortunate that we lose much of the performance [gains from the previous PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2892#issuecomment-814416067), but I think we are still coming in slightly ahead, and perhaps this design gives us a path to further improvement.

One thing I am a little concerned about is testing. Do you think we're sufficiently exercising various configurations? 

One thought I have on the converter, is that it's quite difficult to use for testing purposes. It would be nice if there were a way to "statelessly" convert `entry.Entry` to `pdata.Logs`, without having to work around implementation details of the converter. I think this would allow us to write test cases more easily.

**Comment by pmalek-sumo** [comment]
> One thing I am a little concerned about is testing. Do you think we're sufficiently exercising various configurations?

We can easily add those it depends what you have in mind. This package is transitively tested by other packages which use it so this somewhat makes me more confident that it works hence I didn't spend too much time on testing it.

I'm willing to put more effort if there's will to accept it on this (not only) condition :).

> One thought I have on the converter, is that it's quite difficult to use for testing purposes. It would be nice if there were a way to "statelessly" convert `entry.Entry` to `pdata.Logs`, without having to work around implementation details of the converter. I think this would allow us to write test cases more easily.

I believe we can add a thin wrapper `func` on top of `convert()` which would translate to `pdata.Logs` as it used to (somewhat) before and use that in tests.

**Comment by djaglowski** [comment]
@pmalek-sumo Thanks, I think it makes sense to move forward on this approach. It's looking promising.

**Comment by pmalek-sumo** [comment]
@djaglowski Added `Convert()` for stateless scenarios as you've mentioned.

Not sure what sort of tests we'd like to add. As I've mentioned: this being transitively as a direct dependency of e.g. filelogreceiver or syslogreceiver is rather sufficient.

One small point in this subject I'd like to raise is that if we observe testbed tests to fail we can then address it (of course if need be).

**Review by djaglowski — APPROVED** [review]
This looks great to me. Thanks @pmalek-sumo, for all the iterations on this one.

File: internal/stanza/config.go [hunk 2]
```diff
@@ -43,6 +43,10 @@ type ConverterConfig struct {
 	// FlushInterval defines how often to flush the converted and accumulated
 	// log entries.
 	FlushInterval time.Duration `mapstructure:"flush_interval"`
+	// WorkerCount defines how many worker goroutines used for entry.Entry to
+	// log records translation should be spawned.
+	// By default: runtime.NumCPU() / 4 worker are spawned.
```
Comment (bogdandrutu) [inline internal/stanza/config.go:48]: Hopefully min 1 🗡️

Comment (pmalek-sumo) [reply internal/stanza/config.go]: Should be better now 🤞

**Comment by pmalek-sumo** [comment]
Irrelevant `exporter/elasticexporter` tests failed on Windows

**Comment by bogdandrutu** [comment]
Please rebase

**Comment by pmalek-sumo** [comment]
> Please rebase

@bogdandrutu Done. I can see that testbed test `Log10kDPS/syslog-udp` failed but it seems has been randomly failing before this PR 🤷‍♂️

**Comment by djaglowski** [comment]
> Log10kDPS/syslog-udp failed

Just noting that there is an issue for this already: #3095

For now, I'm rerunning tests.

File: receiver/filelogreceiver/go.mod (hunk 1 above)
Comment (bogdandrutu) [reply receiver/filelogreceiver/go.mod]: No longer changed correct?

Comment (pmalek-sumo) [reply receiver/filelogreceiver/go.mod]: Correct, this was resolved on main hence no longer needed here

**Pull request merged**